File: java/starter/acquirer/src/main/java/network/t0/pay/acquirer/internal/GetPaymentQuote.java
package network.t0.pay.acquirer.internal;

import io.grpc.StatusRuntimeException;
import network.t0.pay.proto.tzero.v1.pay.AcquirerServiceGrpc;
import network.t0.pay.proto.tzero.v1.pay.Decimal;
import network.t0.pay.proto.tzero.v1.pay.GetPaymentQuoteRequest;
import network.t0.pay.proto.tzero.v1.pay.GetPaymentQuoteResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.TimeUnit;

/**
 * §3 GetPaymentQuote — prices an upcoming fiat sale against your LP's standing
 * quotes. Fiat-settlement acquirers only; in USDt mode you run your own FX and
 * pass the rate straight to §4.
 *
 * <p>Stateless lookup, no idempotency key: it always returns the current standing
 * quote, so an {@link Outcome.Unknown} here is safe to retry as often as you like.
 * The returned quoteId is what §4 references.
 */
public final class GetPaymentQuote {

    private static final Logger log = LoggerFactory.getLogger(GetPaymentQuote.class);

    /** A quote is only useful while the customer is still standing there. */
    private static final int TIMEOUT_SECONDS = 5;

    public static Outcome<GetPaymentQuoteResponse.Success> fetch(
            AcquirerServiceGrpc.AcquirerServiceBlockingStub t0,
            String localCurrency,
            Decimal localAmount) {

        try {
            // Deadline on every call: the stub handed out by BlockingNetworkClient
            // carries none, so a stalled t-0 blocks this thread indefinitely.
            GetPaymentQuoteResponse response = t0.withDeadlineAfter(TIMEOUT_SECONDS, TimeUnit.SECONDS)
                    .getPaymentQuote(GetPaymentQuoteRequest.newBuilder()
                            .setLocalCurrency(localCurrency)
                            .setLocalAmount(localAmount)
                            .build());

            switch (response.getResultCase()) {
                case SUCCESS -> {
                    GetPaymentQuoteResponse.Success success = response.getSuccess();
                    log.info("Quote {}: {} {} costs {} USDt at rate {}, expires at {}",
                            success.getQuoteId(),
                            Decimals.format(localAmount),
                            localCurrency,
                            Decimals.format(success.getAmountUsdt()),
                            Decimals.format(success.getFxRate()),
                            Times.format(success.getExpiresAt()));
                    return new Outcome.Accepted<>(success);
                }
                case FAILURE -> {
                    // QUOTE_UNAVAILABLE — your LP is not quoting this currency right now.
                    // TODO: Step 2.1 — tell the POS the sale cannot be priced.
                    String reason = response.getFailure().getReason().name();
                    log.warn("No quote for {} {}: {}",
                            Decimals.format(localAmount), localCurrency, reason);
                    return new Outcome.Rejected<>(reason);
                }
                default -> {
                    return new Outcome.Rejected<>("response carried no result variant");
                }
            }
        } catch (StatusRuntimeException e) {
            log.error("GetPaymentQuote failed: {}", e.getStatus());
            return new Outcome.Unknown<>(e.getStatus().toString());
        }
    }

    private GetPaymentQuote() {
    }
}

File: go/starter/acquirer/internal/get_payment_quote.go
package internal

import (
	"context"
	"log"
	"time"

	"connectrpc.com/connect"
	pay "github.com/t-0-network/usdt-pay-sdk/go/sdk/gen/tzero/v1/pay"
	"github.com/t-0-network/usdt-pay-sdk/go/sdk/gen/tzero/v1/pay/acquirer"
	"github.com/t-0-network/usdt-pay-sdk/go/sdk/gen/tzero/v1/pay/acquirer/acquirerconnect"
)

const quoteTimeout = 5 * time.Second

// FetchQuote prices an upcoming fiat sale. Stateless lookup, no idempotency
// key: an Unknown here is safe to retry as often as you like.
func FetchQuote(
	ctx context.Context,
	t0 acquirerconnect.AcquirerServiceClient,
	localCurrency string,
	localAmount *pay.Decimal,
) Outcome[*acquirer.GetPaymentQuoteResponse_Success] {
	ctx, cancel := context.WithTimeout(ctx, quoteTimeout)
	defer cancel()

	resp, err := t0.GetPaymentQuote(ctx, connect.NewRequest(&acquirer.GetPaymentQuoteRequest{
		LocalCurrency: localCurrency,
		LocalAmount:   localAmount,
	}))
	if err != nil {
		log.Printf("GetPaymentQuote failed: %v", err)
		return Unknown[*acquirer.GetPaymentQuoteResponse_Success]{Detail: err.Error()}
	}

	msg := resp.Msg
	switch msg.GetResult().(type) {
	case *acquirer.GetPaymentQuoteResponse_Success_:
		s := msg.GetSuccess()
		log.Printf("Quote %d: %s %s costs %s USDt at rate %s, expires at %s",
			s.GetQuoteId(),
			DecimalToString(localAmount), localCurrency,
			DecimalToString(s.GetSettlementAmount()),
			DecimalToString(s.GetFxRate()),
			FormatTimestamp(s.GetExpiresAt()))
		return Accepted[*acquirer.GetPaymentQuoteResponse_Success]{Payload: s}

	case *acquirer.GetPaymentQuoteResponse_Failure_:
		reason := msg.GetFailure().GetReason().String()
		log.Printf("No quote for %s %s: %s", DecimalToString(localAmount), localCurrency, reason)
		return Rejected[*acquirer.GetPaymentQuoteResponse_Success]{Reason: reason}

	default:
		return Unknown[*acquirer.GetPaymentQuoteResponse_Success]{Detail: "response carried an unrecognised result variant"}
	}
}
